boards/imix/src/test/sha256_test.rs

```diff
@@ -0,0 +1,59 @@
+//! This tests a software SHA256 implementation. To run this test,
+//! add this line to the imix boot sequence:
+//! ```
+//!     test::sha256_test::run_sha256();
+//! ```
+//! This test takes a dynamic deferred call (for callbacks). It tries to
+//! hash 'hello world' and uses Digest::validate to check that the hash
+//! is correct.
+//!
+//! The expected output is
+//! Sha256Test: Verification result: Ok(true)
+//!
+//! This tests whether the SHA-256 hash of the string "hello hello
+//! hello hello hello hello hello hello hello hello hello hello "
+//! hashes correctly. This string is 12 repetitions of "hello ", so is
+//! 72 bytes long. As SHA uses 64-byte/512 bit blocks, this verifies
+//! that multi-block hashes work correctly.
+
+use capsules::sha256::Sha256Software;
+use capsules::test::sha256::TestSha256;
+use kernel::dynamic_deferred_call::DynamicDeferredCall;
+use kernel::static_init;
+
+pub unsafe fn run_sha256(call: &'static DynamicDeferredCall) {
```

Comment:
Can we set this up to match the Tock test infrastructure?

As this isn't board specific (it can run on any board) it also might be worth making it more generic

Reply:
How would you not pass DynamicDeferredCall? I think that's necessary. The issue is it's software but needs to issue callbacks.

Reply:
I'm happy for this test to be generalized to other platforms, but I currently only have an imix, so would prefer to leave that to someone else.

Reply:
You would do it by creating `Sha256Software` outside the test framework, in `main.rs`

That would mean we can use the generic test framework but also expose SHA support to userspace

Reply:
Is there documentation on how to correctly write a test?

Reply:
Yep! There is heaps

https://doc.rust-lang.org/book/ch11-01-writing-tests.html is a good start
and https://os.phil-opp.com/testing/ has good information on an embedded setup

Reply:
Ah, ok. We don't have documentation on how to write one, just examples.

The [KV System](https://github.com/tock/tock/blob/master/capsules/src/test/kv_system.rs) is a good example of implementing a test as a capsule.

Then you can `static_init!()` it inside a Rust test case like this: https://github.com/tock/tock/blob/master/boards/opentitan/src/tests/tickv_test.rs#L32 . Note that you don't have to do it in a `#[test_case]`, you just can

and you use a global in `main.rs` to pass it to the test case, like this: https://github.com/tock/tock/blob/master/boards/opentitan/src/tests/tickv_test.rs#L32

I'm not suggesting you do anything related to QEMU. Just that you allow the use of the generic Rust test infrastructure, that way we can call the tests from a `cargo test` command instead of having to comment/uncomment the tests out. In the case of the IMIX you can still just comment it out, but it then allows other boards to add it as a test case. Which then means it can be run as part of the CI :)

Reply:
It seems like this should be documented if it's a requested change for a testing approach?

Reply:
> Ah, ok. We don't have documentation on how to write one, just examples.
> 
> The [KV System](https://github.com/tock/tock/blob/master/capsules/src/test/kv_system.rs) is a good example of implementing a test as a capsule.

This test seems strange to me -- the capsule test documentation says to use `tickv_test::run_tickv_tests(kvstore)`, but this is an OpenTitan-specific test. It seems like there's a completely separate testing ecosystem; it transitions tests to being `static mut` variables within main.rs, something which the rest of Tock has tried hard to remove. Put another way, you've chosen a different set of tradeoffs in testing than the rest of Tock has; I'm a little way of starting to transition the rest of Tock over to it without discussion with in the core group. So for now, I'd like to keep the testing as is, consistent with the rest of imix.

Reply:
There seems to be some confusion about QEMU or OpenTitan specifics. The `test_case` and `cargo test` infrastructure is a core part of Rust. It isn't specific to QEMU or OpenTitan. For example libtock-rs is also [using it](https://github.com/tock/libtock-rs/blob/master/doc/Testing.md), although in a slightly different way.

`tickv_test::run_tickv_tests()` will work on any board that has flash and TicKV setup. OpenTitan is just the board that it was first implemented on. OpenTitan also isn't the only board using the Rust `test_case` and `cargo test` infrastructure.

Besides which, that is unrelated. You don't need to write a `test_case` and use `cargo test`, you can just comment out the test case if you want. The main ask was just to make the test case generic and not require any arguments. Then we can run the test case on a regularly tested board to ensure there are no regressions in the SHA implementation. This is what we do with the [alarm tests for example](https://github.com/tock/tock/blob/master/boards/opentitan/src/tests/multi_alarm.rs#L28)

Then we can call the test from the OpenTitan test builds. We don't need a software implementation for OT, as it has a hardware version. By running it in the tests we can run the test case on every single PR!! That way we avoid it bit rotting, which is what happens when we comment out the test case and run it when someone thinks about it.

Reply:
> There seems to be some confusion about QEMU or OpenTitan specifics. The `test_case` and `cargo test` infrastructure is a core part of Rust. It isn't specific to QEMU or OpenTitan. For example libtock-rs is also [using it](https://github.com/tock/libtock-rs/blob/master/doc/Testing.md), although in a slightly different way.

Yeah, I am aware that this is a standard Rust mechanism.

> `tickv_test::run_tickv_tests()` will work on any board that has flash and TicKV setup. OpenTitan is just the board that it was first implemented on. OpenTitan also isn't the only board using the Rust `test_case` and `cargo test` infrastructure.

Then why is it in `boards/opentitan`? Note line 4:

```rust
use crate::{SIPHASH, TICKV};
```

This requires there be a `static mut`, as in `boards/opentitan/src/main.rs`:

```rust
static mut SIPHASH: Option<&capsules::sip_hash::SipHasher24<'static>> = None;
```

If I follow the approach you've taken to implementing an argument-free test, it requires adding `static mut` fields to `main.rs`. Since other platforms don't do this, it means the test is only available for OT. Is this what you mean? So other boards don't use these global Options, but since OT does, then it's possible to run a software regression?

On one hand, I found that these tests are not particularly useful because they don't use hardware; they can only test pure software. On the other this SHA256 implementation is pure software, so it seems like it's actually a good fit for CI.

> Then we can call the test from the OpenTitan test builds. We don't need a software implementation for OT, as it has a hardware version. By running it in the tests we can run the test case on every single PR!! That way we avoid it bit rotting, which is what happens when we comment out the test case and run it when someone thinks about it.

Given my struggle in getting the current OT system up and running, I think "we" is a bit of a stretch here. :) But it seems like there are two separate points here, one is CI checks, the other is running tests on hardware?

Reply:
It seems like the `static mut` is causing confusion. I'll try and explain it.

So the [Rust tests](https://doc.rust-lang.org/book/ch11-01-writing-tests.html) don't take arguments. Which means we somehow need to pass the already created drivers or capsules to the test case. One option would be to create the driver or capsule in the test case, but this isn't great as it deviates from the non-testing setup and results in duplication.

So the solution we use in OpenTitan and other Tock boards is to use a `static mut` to pass capsules and drivers to test cases. This is the same thing all Tock boards do with `CHIP`, `PROCESSES` and `PROCESS_PRINTER` for the panic IO. It's not great, but it's not that awful.

If you aren't using Rust's testing infrastructure, which the IMIX isn't, then you don't really need to worry about this. You could copy the entire `test_case` into `main()` and then just call it directly. No `static mut` required!

What this means is that we now has a generic test case, that can be called via commented out code or the Rust test infrastructure. It's much more generic!

To answer your other questions:

> Then why is it in `boards/opentitan`? Note line 4:

We use it in OpenTitan, but it's not OpenTitan specific

> This requires there be a `static mut`, as in `boards/opentitan/src/main.rs`:

Yes, it does if you want to pass it to a `test_case`, if you want to call it in `main()` you don't need a `static mut`

> it requires adding `static mut` fields to `main.rs`

Only if you want to call it from outside `main()`. If you want to call the test directly from `main()` you don't need to do this

> Since other platforms don't do this, it means the test is only available for OT. Is this what you mean? So other boards don't use these global Options, but since OT does, then it's possible to run a software regression?

I'm not clear what you mean. The software regression can be called directly from `main()` on any board (obviously we need flash though). As well as that other in tree boards use the Rust test infrastructure already. You are correct that only OT does this for the KV tests though.

> On one hand, I found that these tests are not particularly useful in CI because they don't use hardware; they can only test pure software

Agreed, hopefully soon-ish we can run Tock on Verilator in the CI! Which means we are testing hardware as well!

> On the other this SHA256 implementation is pure software, so it seems like it's actually a good fit for CI.

Exactly!

> Given my struggle in getting the current OT system up and running, I think "we" is a bit of a stretch here. :) But it seems like there are two separate points here, one is CI checks, the other is running tests on hardware?

That is strange. There are a few Tock people using OT on Tock, so I'm surprised your are having so many issues. Maybe file a bug case in Tock for the specific problems and we can try and update the docs to address them?

For this PR I don't think we need to worry about hardware. The main goal should just be allowing the test to be used with the Rust test infrastructure. As an extra bonus actually running the test on OT would be awesome!

Reply:
I understand how Rust tests work. But I think some of your claims don't make sense. The test isn't OpenTitan specific, but it's in the OpenTitan boards directory. The core group agreed to remove `static mut` globals from main.rs a long while ago, and this new testing approach re-introduces them. Unless of course you call things from main, but this is not what the tests you've written (or are suggesting I copy) do.

What are the other boards that are using these tests by declaring `static mut` globals in main.rs?

Reply:
I incorporated a test into OpenTitan that follows the same approach used there.